File: leetcode/Algorithms/47.PermutationsII/Solution.java
import java.util.*;

public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        if (nums.length == 0) return new ArrayList<>();
        Queue<List<Integer>> queue = new LinkedList<>();
        queue.add(Collections.singletonList(nums[0]));
        for (int i = 1; i < nums.length; i++) {
            int queueSize = queue.size();
            while (queueSize-- > 0) {
                List<Integer> record = queue.poll();
                for (int j = 0; j <= record.size(); j++) {
                    boolean shouldBreak = j == record.size() || nums[i] == record.get(j);
                    List<Integer> tmp = new ArrayList<>(record);
                    tmp.add(j, nums[i]);
                    queue.add(tmp);
                    if (shouldBreak) break;
                }
            }
        }
        return new ArrayList<>(queue);
    }
}

File: leetcode/Algorithms/47.PermutationsII/Solution.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        result = []
        self.helper(nums, [], result)
        return result

    def helper(self, nums, tmp, result):
        if len(nums) == 1:
            tmp.append(nums[0])
            result.append(tmp)
            return

        unique = set(nums)
        for cursor in unique:
            next_nums, next_tmp = nums[:], tmp[:]
            next_nums.remove(cursor)
            next_tmp.append(cursor)
            self.helper(next_nums, next_tmp, result)

